src/components/lib/kernel/cos_kernel_api.c

```diff
@@ -717,6 +703,104 @@ __page_bump_alloc(struct cos_compinfo *ci, size_t sz, size_t align)
 	return heap_vaddr;
 }
 
+void
+missing_captbl_node_expand(struct cos_compinfo *ci)
```

Comment:
What is the problem this function trying to solve? Is it solving the problem when CAPTBL_FREE_CAP > CAPTBL_EXPAND_SZ the second captabl has not been allocated?

Reply:
This is too complicated as far as I'm concerned.

Reply:
I forgot the full background here. But generally it relates to that I couldn't allocate sinv cap in the past with more cores, thus I added some checking facilities in the kernel and user space to make sure I have cap for the sinv cap. This is needed I think.

Reply:
I'm worried that this looks like it might be solving a bug for your specific scenario, but not for the general case. And it might make that general case more challenging to debug when problems do happen.

I certainly am not questioning if this is solving a bug. I'm sure it is. But I worry that it is just hiding the symptom of a bug in a specific instance. Do you have any recollection if this is a "final solution" to a bug, or just a "workaround"?